zerver/tests/test_auth_backends.py

```diff
@@ -962,3 +990,290 @@ def test_login_success_under_subdomains(self):
+                AUTH_LDAP_USER_DN_TEMPLATE='uid=%(user)s,ou=users,dc=zulip,dc=com'):
+            with self.assertRaisesRegexp(self.mock_ldap.INVALID_CREDENTIALS,
+                                         'uid=hamlet,ou=users,dc=zulip,dc=com:wrong'):
+                self.backend.authenticate('hamlet@zulip.com', 'wrong')
+
+    def test_login_failure_due_to_nonexistent_user(self):
+        # type: () -> None
+        self.mock_ldap.directory = {
+            'uid=nonexitent,ou=users,dc=zulip,dc=com': {
+                'userPassword': 'testing'
+            }
+        }
+        with self.settings(
+                LDAP_APPEND_DOMAIN='zulip.com',
+                AUTH_LDAP_BIND_PASSWORD='',
+                AUTH_LDAP_USER_DN_TEMPLATE='uid=%(user)s,ou=users,dc=zulip,dc=com'):
+            with self.assertRaisesRegexp(self.mock_ldap.INVALID_CREDENTIALS,
+                                         'uid=nonexistent,ou=users,dc=zulip,dc=com:testing'):
+                self.backend.authenticate('nonexistent@zulip.com', 'testing')
+
+    def test_ldap_permissions(self):
+        # type: () -> None
+        backend = self.backend
+        self.assertFalse(backend.has_perm(None, None))
+        self.assertFalse(backend.has_module_perms(None, None))
+        self.assertTrue(backend.get_all_permissions(None, None) == set())
+        self.assertTrue(backend.get_group_permissions(None, None) == set())
+
+    def test_django_to_ldap_username(self):
+        # type: () -> None
+        backend = self.backend
+        with self.settings(LDAP_APPEND_DOMAIN='zulip.com'):
+            username = backend.django_to_ldap_username('"hamlet@test"@zulip.com')
+            self.assertEqual(username, '"hamlet@test"')
+
+    def test_ldap_to_django_username(self):
+        # type: () -> None
+        backend = self.backend
+        with self.settings(LDAP_APPEND_DOMAIN='zulip.com'):
+            username = backend.ldap_to_django_username('"hamlet@test"')
+            self.assertEqual(username, '"hamlet@test"@zulip.com')
+
+    def test_get_or_create_user_when_user_exists(self):
+        # type: () -> None
+        class _LDAPUser(object):
+            attrs = {'fn': ['Full Name'], 'sn': ['Short Name']}
+
+        backend = self.backend
+        email = 'hamlet@zulip.com'
+        user_profile, created = backend.get_or_create_user(email, _LDAPUser())
+        self.assertFalse(created)
+        self.assertEqual(user_profile.email, email)
+
+    def test_get_or_create_user_when_user_does_not_exist(self):
+        # type: () -> None
+        class _LDAPUser(object):
+            attrs = {'fn': ['Full Name'], 'sn': ['Short Name']}
+
+        ldap_user_attr_map = {'full_name': 'fn', 'short_name': 'sn'}
+
+        with self.settings(AUTH_LDAP_USER_ATTR_MAP=ldap_user_attr_map):
+            backend = self.backend
+            email = 'nonexisting@zulip.com'
+            user_profile, created = backend.get_or_create_user(email, _LDAPUser())
+            self.assertFalse(created)
```

Comment:
Wait, shouldn't created by True here?  Or I guess it's false, because the user isn't in a mock directory, but can we add a test with the user in the mock directory that creates the user successfully here?